R/auditClassicalNumberBunching.R

```diff
@@ -0,0 +1,548 @@
+      p <- p + ggrepel::geom_text_repel(ggplot2::aes(label = x, x = x, y = y), data = maxData, vjust = 1, size = 5)
+    }
+    
+    jfaTheme <- ggplot2::theme(panel.grid.major.y = ggplot2::element_line(color = "#cbcbcb", size = 0.5))
+    
+    p <- jaspGraphs::themeJasp(p, legend.position = "none") + jfaTheme
+    
+    plot$plotObject <- p
+  }
+  
+  if(options[["explanatoryText"]]){
+    
+    plotText <- createJaspHtml(gettextf("<b>Figure %i:</b> Histogram of the individual values in the data set.", jaspResults[["figNumber"]]$object), "p")
+    
+    plotText$position <- positionInContainer + 1
+    plotText$dependOn(optionsFromObject = numberBunchingContainer[["numberBunchingHistogram"]])
+    numberBunchingContainer[["numberBunchingHistogramText"]] <- plotText
+  }
+}
+
+.jfaNumberBunchingAddConclusion <- function(options, numberBunchingContainer, jaspResults, ready, position){
+  
+  if(!is.null(jaspResults[["conclusionContainer"]]) || !ready || !options[["explanatoryText"]])
+    return()
+  
+  conclusionContainer <- createJaspContainer(title= gettext("<u>Conclusion</u>"))
+  conclusionContainer$position <- position
+  conclusionContainer$dependOn(options = c("values",
+                                           "confidence",
+                                           "shuffle",
+                                           "explanatoryText",
+                                           "noSamples",
+                                           "seed"))
+  
+  confidenceLabel <- paste0(round(options[["confidence"]] * 100, 2), "%")
+  
+  state <- .jfaNumberBunchingState(dataset = NULL, options, jaspResults, ready)
+  
+  approve <- state[["pvalueAvgFrequency"]] >= (1 - options[["confidence"]])
+  
+  conclusion <- ifelse(approve, no = gettext("is rejected"), yes = gettext("is not rejected"))
+  
+  pvalue <- round(state[["pvalueAvgFrequency"]], 3)
+  if(pvalue < 0.001)
+    pvalue <- "< .001"
+  pvalue <- ifelse(approve, no = paste0(pvalue, " < \u03B1"), yes = paste0(pvalue, " >= \u03B1"))
+  
+  conclusionText <- gettextf("The <i>p</i> value is %1$s and the null hypothesis that the data do not contain an unexpected amount of repeated values <b>%2$s</b>.", pvalue, conclusion)
+  
+  conclusionContainer[["conclusionParagraph"]] <- createJaspHtml(conclusionText, "p")
+  conclusionContainer[["conclusionParagraph"]]$position <- 1
+  conclusionContainer$dependOn(options = c("explanatoryText", 
+                                           "confidence",
+                                           "values",
+                                           "digits",
+                                           "distribution"))
+  
+  jaspResults[["conclusionContainer"]] <- conclusionContainer
+}
+
+.jfaEntropy <- function(x){
+  frequencies <- as.numeric(table(x))
+  # Probailities instead of frequencies
+  prob <- frequencies / length(x)
+  # Compute sum(p*log(p))
+  entropy <- -sum(prob*t(log(prob)))
+  return(entropy) 
```

Comment:
```suggestion
  prob <- frequencies / sum(frequencies)
  logProb <- log(prob)
  
  # Compute sum(p*log(p))
  entropy <- -sum(ifelse(!is.infinite(logProb), prob*logProb, 0))
```
So I am not sure whether or not `x` can be a factor and whether it can contain `NA`s, but if it can, it's probably a good idea to catch if some of the options has been observed zero times (i.e., avoid multiplying `0` and `-Inf`) and do not assume that `length(x) == n` .
```